aleph/logic/xref.py

```diff
@@ -272,12 +272,38 @@ def xref_collection(collection):
     )
     log.info(f"[{collection}] Clearing previous xref state....")
     delete_xref(collection, sync=True)
-    delete_entities(collection.id, origin=ORIGIN, sync=True)
+    delete_entities(collection.id, origin='xref', sync=True)
+
+    batch = []
+    batch_size = int(SETTINGS.XREF_SCROLL_SIZE)  # Ensure batch_size is an integer
+    for entity in iter_entities(collection_id=collection.id):
+        batch.append(entity['id'])
+        if len(batch) >= batch_size:
+            enqueue_xref_batch(collection, batch)
+            batch = []
+
+    if batch:
+        enqueue_xref_batch(collection, batch)
+
+    log.info(f"[{collection}] Xref batches enqueued, processing will continue in the background.")
+    refresh_collection(collection.id)
+
+
+def enqueue_xref_batch(collection, entity_ids):
+    """Enqueue a task to process a batch of entities for cross-referencing."""
+    queue_task(collection, SETTINGS.STAGE_XREF_BATCH, entity_ids=entity_ids)
+
+
+def process_xref_batch(collection, entity_ids):
+    """Process a batch of entities for cross-referencing."""
+    log.info(f"[{collection}] Processing xref batch with {len(entity_ids)} entities")
     index_matches(collection, _query_entities(collection))
     index_matches(collection, _query_mentions(collection))
-    log.info(f"[{collection}] Xref done, re-indexing to reify mentions...")
-    reindex_collection(collection, sync=False)
 
+    # TODO: I just realized I had accidentally removed these lines, which is probably why it was not reindexing/mapping.
+    # I have to test this first before I can confidently add these back. Currently, at least there is nothing breaking catastrophically.
+    # log.info(f"[{collection}] Xref done, re-indexing to reify mentions...")
+    # reindex_collection(collection, sync=False)
```

Comment:
Oh, and in case you’ve never had to cross-reference mentions from documents before, here’s how you can test it:

* Create Investigation A.
* Create a few person entities with common names in Investigation A.
* Create Investigation B.
* Create a PDF or plaintext document that contains the names of the entities in Collection A. Upload the document to Investigation B and make sure that Aleph recognized the names in the document.
* Run cross-referencing in Investigation B. Once it has completed, you should see matches for the names in the document.

Reply:
I would imagine that we need to reindex per batch now, otherwise the batches will become dependent on each other again for success/completion? I have to take another look at this as well.